crypto/cms/cms_lib.c

```diff
@@ -74,10 +74,13 @@ CMS_ContentInfo *CMS_ContentInfo_new(void)
 void CMS_ContentInfo_free(CMS_ContentInfo *cms)
 {
     if (cms != NULL) {
-        CMS_EncryptedContentInfo *ec = ossl_cms_get0_env_enc_content(cms);
-
-        if (ec != NULL)
-            OPENSSL_clear_free(ec->key, ec->keylen);
+        /* Check content type is enveloped data to avoid CMS_R_CONTENT_TYPE_NOT_ENVELOPED_DATA error */
```

Comment:
Maybe it would be cleaner to call a new  function here..
The function could live in cms_env.c (with header cms_local.h)
e.g. ossl_cms_env_enc_content_free()
which could then use cms_get_enveloped_type()

Reply:
````
static int cms_get_enveloped_type_simple(const CMS_ContentInfo *cms)
{
    int nid = OBJ_obj2nid(cms->contentType);

    switch (nid) {
    case NID_pkcs7_enveloped:
        return CMS_ENVELOPED_STANDARD;

    case NID_id_smime_ct_authEnvelopedData:
        return CMS_ENVELOPED_AUTH;

    default:
        return 0;
    }
}

static int cms_get_enveloped_type(const CMS_ContentInfo *cms)
{
    int ret = cms_get_enveloped_type_simple(cms);

    if (ret == 0)
        ERR_raise(ERR_LIB_CMS, CMS_R_CONTENT_TYPE_NOT_ENVELOPED_DATA);
    return ret;
}
````

Reply:
````
void ossl_cms_env_enc_content_free(const CMS_ContentInfo *cinf)
{
    if (cms_get_enveloped_type_simple(cinf) != 0)
    {
            CMS_EncryptedContentInfo *ec = ossl_cms_get0_env_enc_content(cms);

            if (ec != NULL)
                OPENSSL_clear_free(ec->key, ec->keylen);
    }
}
````